### ISSUE TITLE
sfetch not working on Emscripten in custom bindings

### ISSUE THREAD

**Issue by troyedwardsjr** [floooh/sokol-rust#21]
@floooh Don't wish to bother you, but not sure where else to ask this question. Would love an official sokol Discord server. 

sfetch is not working on Emscripten in my custom Rust bindings.

Having one async cross-platform abstraction already built for fetch would be very nice, so I'd like to use it if possible. All other sokol bindings are working.

sfetch is not working on Emscripten for me from my custom Rust bindings.

All the symbols are resolving properly and I have `SOKOL_IMPL` defined. What could be the issue?

I call `sfetch_dowork()` on the update callback for sokol / Emscripten loop / requestAnimationFrame().
I initialize it on the init setup callback like this:
```rust
        let mut sfetch_desc: sfetch_desc_t = core::mem::MaybeUninit::zeroed().assume_init();
        sfetch_desc.max_requests = 3;
        sfetch_desc.num_channels = 1;
        sfetch_desc.num_lanes = 1;
        sfetch_desc.logger.func = Some(sokol::log::slog_func);     
        sfetch_setup(&mut sfetch_desc);
        if !sfetch_valid() {
            panic!("sfetch is not valid");
        }
 ```
This all works. sfetch is valid.

Then when I call this everything freezes:
```rust
    #[repr(C)]
    struct Buffers {
        atlas: [u8; 4 * 1024],
        skeleton: [u8; 128 * 1024],
        image: [u8; 512 * 1024]
    }
    let buffers = Buffers {
        atlas: [0; 4 * 1024],
        skeleton: [0; 128 * 1024],
        image: [0; 512 * 1024]
    };
    unsafe {
        let mut sfetch_request: sfetch_request_t = core::mem::MaybeUninit::zeroed().assume_init();
        sfetch_request.path = "assets/character.png\0".as_ptr() as *const i8;
        sfetch_request.channel = 0;
        sfetch_request.buffer = sfetch_range_t {
            ptr:  &buffers.image as *const u8 as *const c_void,
            size: 512 * 1024 
        };
        sfetch_request.callback = Some(atlas_data_loaded);
        sfetch_send(&sfetch_request);
    }
```

```
PS C:\Users\troye\dev\toxoid\toxoid> emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.49 (04a0cad4d4c9f5d62876821274d78cd0a52427af)
clang version 18.0.0 (https://github.com/llvm/llvm-project 269685545e439ad050b67740533c59f965cae955)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: C:\Users\troye\dev\bin\emsdk\upstream\bin
```

I know that Emscripten is single-threaded by default unless you enable pthreads. But I thought this was using the `EM_JS` macro to call things asynchronously from the JS browser host. Is there something I'm missing? Why does it freeze at `sfetch_send(&sfetch_request);` with no error logs?

**Comment by floooh** [comment]
Are there any messages in the browser's Javascript console? The web app appearing to "freeze" is usually just a symptom that the app has hit a Javascript exception.

If it's a real freeze (like in and infinite loop) the browser would kill the tab after at most 30 seconds or so and you'd get an "Aw snap!" screen from Chrome.

**Comment by troyedwardsjr** [comment]
@floooh Sorry let me rephrase, not a browser freeze, a stop/block in the execution of the program where nothing prints. Oddly nothing before it prints either when this code is included, but the rest of sfetch is fine, which might represent a more complex issue between the compiling / linking steps from C -> Rust -> Emscripten. So I thought it might be some kind of symbol resolution issue, but typically Emscripten prints those as an exception as well. So I'm not sure what's going on.

**Comment by floooh** [comment]
Hmm weird indeed.

sokol_fetch.h isn't part of the other bindings currently, so I haven't paid much attention yet to 'binding compatibility' (but I can't think of any obvious potential problems). There's quite a bit of internal complexity in sokol_fetch.h (e.g. use of threading on native platforms). The Javascript backend is comparatively simple though, it's just using regular `Promise.then()` async calls on the browser thread.

Looking at `sfetch_send()` (https://github.com/floooh/sokol/blob/17b399b8eeb6d1db4fe0e78fbea3b1a4184470be/sokol_fetch.h#L2709-L2733) there's not much going on there, and expecially nothing backend- or threading-specific (no calls into Javascript for instance).

There are also no loops so a "freeze" is still mysterious.

**Comment by troyedwardsjr** [comment]
Isolated the fetch code into its codebase and found the issue, and for whatever reason in my more complicated build setup, the error did not appear. But in this smaller setup, it did appear.

There were two issues:

1. My code was probably not blocking execution for fetch before the sokol init callback (sfetch_setup) was called.
2. `thread has overflowed its stack` was occurring because I stored the buffer (512 * 1024) in the stack instead of the heap with a vec, which I'm now doing and passing in a raw pointer from the vec. However, I am wondering how I should determine the size of the file before preallocating a buffer. 

All clear!

Thanks again for the sanity check @floooh!

**Comment by floooh** [comment]
> I am wondering how I should determine the size of the file before preallocating a buffer.

Yeah that's a known limitation unfortunately, mainly caused by the web platform not being able to provide an overall uncompressed file size before the download has finished.

In a game-style application I would solve this via a separate table-of-content-file which contains the size and maybe a checksum for each asset file.
